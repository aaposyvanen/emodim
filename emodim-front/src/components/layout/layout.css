.layout {
    display: flex;
    flex-flow: column;
    align-items: center;
    width: 100%;
    height: 100%;
    min-height: 100vh;
}
