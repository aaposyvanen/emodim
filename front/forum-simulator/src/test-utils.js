import React from 'react'
import { render as rtlRender } from '@testing-library/react'
import { createStore } from 'redux'
import { Provider } from 'react-redux'

import createRootReducer from './reducer'

// This can be used instead of rtlRender to easily render connected (Redux) components.
// Can be especially useful when children of the tested component are connected.
// There might be a better way, (e.g. shallow rendering) but at least this works.
// Inspiration for this approach:
// https://redux.js.org/recipes/writing-tests#connected-components

function render(
    ui,
    {
        initialState,
        store = createStore(createRootReducer, initialState),
        ...renderOptions
    } = {}
) {
    function Wrapper({ children }) {
        return <Provider store={store}>{children}</Provider>
    }
    return rtlRender(ui, { wrapper: Wrapper, ...renderOptions })
}

// re-export everything
export * from '@testing-library/react'
// override render method
export { render }
