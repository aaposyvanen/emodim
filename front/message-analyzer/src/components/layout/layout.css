.layout {
    background-color: #90c0e0;
    width: 100%;
    height: 100%;
    min-height: 100vh;
}

.forum-view {
    display: flex;
    flex-direction: column;
    align-items: center;
}
