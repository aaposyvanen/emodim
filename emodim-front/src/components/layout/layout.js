import React from "react";

import { useSelector } from "react-redux";

import ForumView from "../forumView/forumView";
import UserSelection from "../userSelection/userSelection";

import "./layout.css";

/*
 * Shows username form if no username has been chosen yet 
 */
export const Layout = () => {
    const username = useSelector(state => state.userReducer.username);
    const usernameIsChosen = username !== "";    

    return (
        <div className="layout">
            {usernameIsChosen
                ? <ForumView />
                : <UserSelection />}
        </div>
    );
}

export default Layout;
