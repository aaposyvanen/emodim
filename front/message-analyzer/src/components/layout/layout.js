import React, { useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import _ from "lodash";
import {
    updateAvailableRawThreads,
    updateCurrentRawThread,
    updateCurrentRawThreadIndex
} from "../../actions/rawDataActions";
import { updateCurrentThread } from "../../actions/threadActions";
import json from "../../testData/threadData_s24_04.json";
import Thread from "../thread/thread";
import ThreadChanger from "../threadChanger/threadChanger";
import "./layout.css";

export const Layout = () => {

    const dispatch = useDispatch();

    useEffect(() => {
        dispatch(updateAvailableRawThreads(json));
    }, [dispatch]);

    const availableRawThreads = useSelector(state => state.rawDataReducer.availableThreads);
    const currentRawThreadIndex = useSelector(state => state.rawDataReducer.currentIndex);
    const currentRawThread = useSelector(state => state.rawDataReducer.currentThread)

    useEffect(() => {
        if (currentRawThread && currentRawThread.comments) {
            const [comments, responses] = separateResponsesFromComments(currentRawThread.comments)
            const commentsWithResponses = moveResponsesToTheirParents(comments, responses);
            const reverseOrderedComments = _.reverse(commentsWithResponses); // newest comments are shown first
            currentRawThread.comments = reverseOrderedComments;
            dispatch(updateCurrentThread(currentRawThread));
        }
    }, [currentRawThread, dispatch]);

    const separateResponsesFromComments = (rawComments) => {
        const groupedComments = _.groupBy(rawComments, comment => {
            return comment.commentMetadata.parent_comment_id === "0";
        });

        const comments = groupedComments["true"];
        const responses = groupedComments["false"];
        return [comments, responses];
    }

    const moveResponsesToTheirParents = (comments, responses) => {
        const messageChain = [];

        _.forEach(comments, comment => {
            // Find the children of each comment
            const children = _.filter(responses, response => {
                const parentId = response.commentMetadata.parent_comment_id;
                const commentId = comment.commentMetadata.id.split(":")[1];
                return parentId === commentId
            });

            if (!_.isEmpty(children)) {
                if (_.has(comment, "children")) {
                    _.concat(comment.children, children)
                } else {
                    comment.children = children;
                }
            }
            messageChain.push(comment);
        });
        return messageChain;
    }

    useEffect(() => {
        if (!_.isEmpty(availableRawThreads)) {
            dispatch(updateCurrentRawThread(availableRawThreads[currentRawThreadIndex]));
        }
    }, [availableRawThreads, currentRawThreadIndex, dispatch]);

    return (
        <div className="layout">
            <div className="forum-view">
                <ThreadChanger
                    currentIndex={currentRawThreadIndex}
                    updateIndex={(index) => dispatch(updateCurrentRawThreadIndex(index))}
                    maxIndex={availableRawThreads.length - 1}
                />
                <Thread />
            </div>
        </div>
    );
}

export default Layout;
