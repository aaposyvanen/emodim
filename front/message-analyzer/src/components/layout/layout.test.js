import React from "react";
import { render } from "../../test-utils";
import { Layout } from "./layout";

jest.mock("../thread/thread", () => () => "Thread");

test("Layout renders correctly", () => {

    const { container } = render(<Layout />);

    expect(container).toMatchSnapshot();
});
