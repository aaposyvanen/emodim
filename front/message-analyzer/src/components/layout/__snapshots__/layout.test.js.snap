// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Layout renders correctly 1`] = `
<div>
  <div
    class="layout"
  >
    <div
      class="forum-view"
    >
      <div
        class="thread-changer"
      >
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Prev
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
        <div
          class="current-thread-label"
        >
          1
        </div>
        <button
          class="MuiButtonBase-root MuiButton-root MuiButton-text"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiButton-label"
          >
            Next
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
      Thread
    </div>
  </div>
</div>
`;
